### ISSUE TITLE
turbo-frame loading=lazy loads eagerly in Safari

### ISSUE THREAD

**Issue by jgaskins** [hotwired/turbo#260]
It appears that the `IntersectionObserver` in Safari fires immediately on custom HTML elements. For some apps using lazy turbo frames, this is effectively graceful degradation (which is great!) but there have been a few blog posts written about using lazy turbo frames for infinite scroll ([this one](https://dev.to/andrzejkrzywda/from-rails-scaffold-listing-to-hotwire-infinite-scroll-3273), for example, which links to a couple others).

In this use case, due to this issue in Safari, infinite-scrolling feeds will continue loading forever until the user reaches the bottom of the feed, the server rate-limits them, or their browser crashes.

**Comment by filipemendespi** [comment]
@jgaskins Do you have any solution for this problem? The application that I work on has the same error, even on pages that have the turbo-drive disabled.

**Comment by jgaskins** [comment]
@filipemendespi None that I’ve tested, unfortunately. Turbo Frames being implemented in terms of custom elements is a fundamental design decision. As a workaround, you might be able to attach an `IntersectionObserver` to a standard element with a Stimulus controller and have _that_ observer attach the `src` attribute to the `turbo-frame` element. I don’t know if that will work, but the basic idea is this:

On page load:

```html
<turbo-frame id=“my-frame”>
  <div id=“my-frame-observer”></div>
</turbo-frame>
```

When `my-frame-observer` hits the viewport:

```html
<turbo-frame id=“my-frame” src=“/my_data”>
</turbo-frame>
```

I _believe_ this should kick off the request as expected, but I haven’t tested it yet. Note that this is just the HTML part of it and it doesn’t include any Stimulus wiring or even vanilla JS event handlers you’d have to implement. If I can test it out some time this week, I’ll update this thread.

**Comment by filipemendespi** [comment]
as I said, I have **hotwire** enabled only in some views of the application, interestingly this error occurs on pages where the turbo-drive is disabled and I do not use anything related to the turbo, just load it because I have the includes in the application layout.
I don't think this solution could help me, considering that the pages that fail do not have `turbo-frame`.

**Comment by jgaskins** [comment]
Then I think your question has landed in the wrong place. This issue is about `<turbo-frame loading=lazy>`.

**Comment by tleish** [comment]
In my project `turbo-frame loading="lazy"` works on Safari as expected.  Does someone have a github repo where it's not working on Safari?

**Comment by jgaskins** [comment]
Huh, just checked and it’s working for me now, too. I wonder if this was fixed in the latest Safari release? Because I did reproduce this (worked as expected in Firefox but loaded the whole list in Safari without scrolling) and was even able to replicate it with just an `IntersectionObserver` watching a custom element that was outside the viewport.

Either way, it looks like it’s fixed in Safari so a workaround shouldn’t be needed. I’ll go ahead and close.
